src/widgets/list.rs

```diff
@@ -142,31 +165,52 @@ impl<'a> List<'a> {
         let mut start = offset;
         let mut end = offset;
         let mut height = 0;
+        // calculate which items will be showing starting from offset and
+        // filling only the items that fit in the available space
         for item in self.items.iter().skip(offset) {
+            // don't include the padding in the max height calculation
+            // as we don't need to render the last item's padding
             if height + item.height() > max_height {
                 break;
             }
-            height += item.height();
+            height += item.height() + self.padding;
             end += 1;
         }
 
+        // if the selected item is after the range of items that will be
+        // showing, then we need to adjust the start and end bounds to
+        // reflect this. We add one item at the end, and remove 0 or more
+        // items from the start to ensure that the height is less than or
+        // equal to the max height.
         let selected = selected.unwrap_or(0).min(self.items.len() - 1);
         while selected >= end {
-            height = height.saturating_add(self.items[end].height());
+            height = height.saturating_add(self.items[end].height() + self.padding);
             end += 1;
             while height > max_height {
-                height = height.saturating_sub(self.items[start].height());
+                height = height.saturating_sub(self.items[start].height() + self.padding);
                 start += 1;
             }
         }
+
+        // here we do the same thing as above, but in the opposite direction
+        // to ensure that the start and end bounds are correct. We remove one
+        // item from the start, and add 0 or more items to the end to ensure
+        // that the height is less than or equal to the max height.
         while selected < start {
```

Comment:
this `while` can be replaced by a simple if at the beginning: here the complete code of `get_items_bounds`.

```
fn get_items_bounds(
    &self,
    selected: Option<usize>,
    offset: usize,
    max_height: usize,
) -> (usize, usize) {
    let mut start = offset.min(self.items.len().saturating_sub(1));
    if let Some(selected) = selected {
        start = start.min(selected);
    }
    let mut end = start;
    let mut height = 0;
    // calculate which items will be showing starting from offset and
    // filling only the items that fit in the available space
    for item in self.items.iter().skip(start) {
        // don't include the padding in the max height calculation
        // as we don't need to render the last item's padding
        if height + item.height() > max_height {
            break;
        }
        height += item.height() + self.padding;
        end += 1;
    }

    // if the selected item is after the range of items that will be
    // showing, then we need to adjust the start and end bounds to
    // reflect this. We add one item at the end, and remove 0 or more
    // items from the start to ensure that the height is less than or
    // equal to the max height.
    let selected = selected.unwrap_or(0).min(self.items.len() - 1);
    while selected >= end {
        height = height.saturating_add(self.items[end].height() + self.padding);
        end += 1;
        while height > max_height {
            height = height.saturating_sub(self.items[start].height() + self.padding);
            start += 1;
        }
    }

    if !self.truncate_last_item && height < max_height {
        // if the height is less than the max height, then we need to
        // adjust the end bound to include the last item.
        end += 1;
    }

    (start, end)
}
```